analyses/malawi/flooding_trigger/mwi-trigger.md

```diff
@@ -0,0 +1,222 @@
+    df_rp = (observations.to_dataframe()
+                 .rename(columns={station: 'discharge'})
+                 .resample(rule='A', kind='period')
+                 .max() 
+                 .sort_values(by='discharge', ascending=False)
+                )
+    df_rp["year"] = df_rp.index.year
+     
+    n = len(df_rp)
+    df_rp['rank'] = np.arange(n) + 1
+    df_rp['exceedance_probability'] = df_rp['rank'] / (n+1)
+    df_rp['rp'] = 1 / df_rp['exceedance_probability']
+    return interp1d(df_rp['rp'], df_rp['discharge'])
+
+rp_dict = {}
+
+for station in STATIONS:
+    f_rp = get_return_period_function(da_glofas_reanalysis[station], station)
+    rp_dict[station] = {}
+    for year in [1.5, 2, 3, 4, 5, 10, 20]:
+        val = 10*np.round(f_rp(year) / 10)
+        rp_dict[station][year] = val
+
+df_rps = pd.DataFrame(rp_dict)
+```
+
+### Overview of historical discharge
+
+```python
+# Return periods to focus on, with display colours
+rps = {
+    3: '#32a852',
+    5: '#9c2788'
+}
+
+for station in STATIONS: 
+    da_plt = da_glofas_reanalysis[station].sel(time=slice('1999-01-01','2020-12-31'))
+    df_flood_mvac = df_mvac_flood_ta[df_mvac_flood_ta['name']==station]
+
+    fig, ax = plt.subplots()
+    da_plt.plot(x='time', add_legend=True, ax=ax)
+    ax.set_title(f'Historical streamflow at {station}')
+    ax.set_xlabel("Date")
+    ax.set_ylabel('Discharge [m$^3$ s$^{-1}$]')
+
+    for i in range(0,len(df_floodscan_event['start_date'])):
+        ax.axvspan(np.datetime64(df_floodscan_event['start_date'][i]), np.datetime64(df_floodscan_event['end_date'][i]), alpha=0.5, color='#FE5E1E')
+    
+    for key, value in rps.items():
+        ax.axhline(rp_dict[station][key], 0, 1, color=value, label=f'{str(key)} return period')
+        
+    ax.legend()
+    
+    plt.savefig(PLOT_DIR / f'{station}_historical_discharge_glofas_overview_rps.png')
+```
+
+### Identifying glofas events
+
+```python
+def get_groups_above_threshold(observations, threshold):
+    return np.where(np.diff(np.hstack(([False],
+                                           observations > threshold,
+                                           [False]))))[0].reshape(-1, 2)
+
+def get_detection_stats(df_glofas, df_impact, buffer=True):
+    TP = 0 
+    FP = 0
+    tot_events = len(df_impact.index)
+    tot_activations = len(df_glofas.index)
+
+    for index, row in df_glofas.iterrows():
+
+        TP_ = False
+        act_dates = np.array(pd.date_range(row['start_date'], row['end_date']))
+
+        for index,row in df_impact.iterrows():
+            if buffer:
+                event_dates = np.array(pd.date_range(row['start_date_buffer'], row['end_date_buffer']))
+            else:
+                event_dates = np.array(pd.date_range(row['start_date'], row['end_date']))
+            if (set(act_dates) & set(event_dates)):
+                TP+=1
+                TP_ = True
+                df_impact = df_impact.drop([index,])
```

Comment:
One edge case where this may not work is if there were two activation dates close together that could be associated with the same event -- the second activation would count as a false positive

Reply:
Ohhh good point, thanks! I should catch this.